squeal-postgresql-ltree/src/Squeal/PostgreSQL/LTree.hs

```diff
@@ -0,0 +1,412 @@
+      . (<> " :: lquery")
+      . escapeQuotedString
+instance IsString
+  (Expression grp lat with db params from (null PGltxtquery)) where
+    fromString
+      = UnsafeExpression
+      . parenthesized
+      . (<> " :: ltxtquery")
+      . escapeQuotedString
+
+-- | Returns subpath of ltree from position start to position end-1 (counting from 0).
+subltree :: '[null PGltree, null 'PGint4, null 'PGint4] ---> null PGltree
+subltree = unsafeFunctionN "subltree"
+
+
+-- | Returns subpath of ltree starting at position offset, with length len.
+-- If offset is negative, subpath starts that far from the end of the path.
+-- If len is negative, leaves that many labels off the end of the path.
+subpath :: '[null PGltree, null 'PGint4, null 'PGint4] ---> null PGltree
+subpath = unsafeFunctionN "subpath"
+
+-- | Returns subpath of ltree starting at position offset,
+-- extending to end of path. If offset is negative,
+-- subpath starts that far from the end of the path.
+subpathEnd :: '[null PGltree, null 'PGint4] ---> null PGltree
+subpathEnd = unsafeFunctionN "subpath"
+
+-- | Returns number of labels in path.
+nlevel :: null PGltree --> null 'PGint4
+nlevel = unsafeFunction "nlevel"
+
+-- | Returns position of first occurrence of b in a, or -1 if not found.
+indexLTree :: '[null PGltree, null PGltree] ---> null 'PGint4
+indexLTree = unsafeFunctionN "index"
+
+-- | Returns position of first occurrence of b in a, or -1 if not found.
+-- The search starts at position offset;
+-- negative offset means start -offset labels from the end of the path.
+indexOffset :: '[null PGltree, null PGltree, null 'PGint4] ---> null 'PGint4
+indexOffset = unsafeFunctionN "index"
+
+-- | Casts text to ltree.
+text2ltree :: null 'PGtext --> null PGltree
+text2ltree = unsafeFunction "text2ltree"
+
+-- | Casts ltree to text.
+ltree2text :: null PGltree --> null 'PGtext
+ltree2text = unsafeFunction "ltree2text"
+
+-- | Computes longest common ancestor of paths in array.
+lca :: null ('PGvararray ('NotNull PGltree)) --> null PGltree
+lca = unsafeFunction "lca"
+
+
+{- |
+`(@>)` Is left argument an ancestor of right (or equal)?
+
+`(<@)` Is left argument a descendant of right (or equal)?
+-}
+instance PGSubset PGltree where
+  (@>) = unsafeBinaryOp "@>"
+  (<@) = unsafeBinaryOp "<@"
+
+-- | Does ltree match lquery?
+(%~) :: Operator (null0 PGltree) (null1 PGlquery) ('Null 'PGbool)
+(%~) = unsafeBinaryOp "~"
+
+-- | Does ltree match lquery?
+(~%) :: Operator (null1 PGlquery) (null0 PGltree) ('Null 'PGbool)
+(~%) = unsafeBinaryOp "~"
+
+-- | Does ltree match any lquery in array?
+(%?) :: Operator
+  (null0 PGltree) (null1 ('PGvararray (null2 PGlquery))) ('Null 'PGbool)
+(%?) = unsafeBinaryOp "?"
+
+-- | Does ltree match any lquery in array?
+(?%) :: Operator
+  (null0 ('PGvararray (null1 PGlquery))) (null2 PGltree) ('Null 'PGbool)
+(?%) = unsafeBinaryOp "?"
+
+-- | Does ltree match ltxtquery?
+(%@) :: Operator (null PGltree) (null PGltxtquery) ('Null 'PGbool)
```

Comment:
should these be `null0` `null1`? Doesn't seem like the nullability has to match

Reply:
yup!